nab/labeler.py

```diff
@@ -206,111 +209,104 @@ def getUserLabels(self):
 
-
-  def combineLabels(self):
-    """Combine windows to create raw labels.
-    This uses the threshold to determine if a particular record should be
-    labeled as 1 or 0; threshold describes the level of agreement
-    between labelers.
+    
+  def combineRawLabels(self):
+    """
+    Combines raw user labels to create set of true anomaly labels; checks 
+    explicitly for common start times in the raw windows.
+    A buffer is used to merge labels that identify the same anomaly. The buffer
+    is half the estimated window size of an anomaly -- approximates an average
+    of two anomalies per dataset, and no window can have >1 anomaly.
+    After merging, a label becomes a true anomaly if it was labeled by a
+    proportion of the users greater than the defined threshold.
     """
     combinedLabels = {}
-
+    labelIndices = {}
+    
     for relativePath, dataSet in self.corpus.dataFiles.iteritems():
-      timestamps = []
-      labels = []
       
-      for _, row in dataSet.data.iterrows():
-        t = row["timestamp"]
-
-        count = 0
-        for user in self.userLabels:
-          if any(t1 <= t <= t2 for [t1,t2] in user.windows[relativePath]):
-            count += 1
-
-        # Label anomalous if count is large enough, as defined by threshold
-        label = int(count >= self.nlabelers * self.threshold and count > 0)
-
-        timestamps.append(t)
-        labels.append(label)
-
-      combinedLabels[relativePath] = pandas.DataFrame({"timestamp":timestamps,
-        "label": labels})
-
+      length = len(dataSet.data)
+      timestamps = dataSet.data["timestamp"]
+      buffer = datetime.timedelta(minutes=round(length*self.windowSize/2))
+      rawWindows = []
+      bucket = []
+      rawAnomalies = []
+      trueAnomalies = []
+      
+      for user in self.userLabels:
+        if user.windows[relativePath]:
+          rawWindows.append(user.windows[relativePath])
+      if not rawWindows:
+          # No labeled anomalies in this dataset
+          combinedLabels[relativePath] = pandas.DataFrame(
+            {"timestamp":timestamps, "label":numpy.zeros(length, dtype=int)})
+          labelIndices[relativePath] = []
+          continue
+      else:
+        times = []
+        # Get all the labeled start times
+        [times.append(t[0]) for timesList in rawWindows for t in timesList]
+      times.sort()
+
+      # Bucket similar timestamps
+      current = None
+      for t in times:
+        if current is None:
+          current = t
+          bucket = [current]
+          continue
+        if (t - current) <= buffer:
+          bucket.append(t)
+        else:
+          rawAnomalies.append(bucket)
+          current = t
+          bucket = [current]
+      if bucket:
+        rawAnomalies.append(bucket)
+
+      # Merge the bucketed timestamps that qualify as true anomalies
+      for bucket in rawAnomalies:
+        if len(bucket) > len(self.userLabels)*self.threshold:
```

Comment:
Should it be `>=` here? If threshold is 50% then we would want 2 out of 4 to pass.